ambry-utils/src/main/java/com.github.ambry.utils/Utils.java

```diff
@@ -912,6 +913,30 @@ public static DataInputStream createDataInputStreamFromBuffer(Object buffer, boo
         .collect(Collectors.toCollection(collectionFactory));
   }
 
+  /**
+   * Partition the input list into a List of smaller sublists, each one limited to the specified batch size.
+   * Method inspired by the Guava utility Lists.partition(List<T> list, int size).
+   * @param inputList the input list to partition.
+   * @param batchSize the maximum size of the returned sublists.
+   * @return the partitioned list of sublists.
+   */
+  public static <T> List<List<T>> partitionList(List<T> inputList, int batchSize) {
+    Objects.requireNonNull(inputList, "Input list cannot be null");
+    if (batchSize < 1) {
+      throw new IllegalArgumentException("Invalid batchSize: " + batchSize);
+    }
+    List<List<T>> partitionedLists = new ArrayList<>();
+    for (int j = 0; j < inputList.size() / batchSize + 1; j++) {
```

Comment:
could this loop be simplified to:
```
for (int start = 0; start < inputList.size(); start += batchSize) {
  int end = Math.min(start + batchSize, inputList.size());
  partitionedLists.add(inputList.subList(start, end));
}
```
I may be missing an edge case here though

Reply:
So I was thinking about this partition thing and it seems that we don't really need a full list and just need an iterable that we can do a for loop on. Something like this could work, I think:
```java
  private static <T> Iterable<List<T>> partitionIterable(List<T> list, int batchSize) {
    return () -> new Iterator<List<T>>() {
      int nextStart = 0;

      @Override
      public boolean hasNext() {
        return nextStart < list.size();
      }

      @Override
      public List<T> next() {
        if (!hasNext()) {
          throw new NoSuchElementException();
        }
        int start = nextStart;
        nextStart += batchSize;
        return list.subList(start, Math.min(start + batchSize, list.size()));
      }
    };
  }
```

Reply:
I'm not sure this is a big improvement.  I prefer to mimic the guava utility.  Once we upgrade to a newer JDK, we can probably retire the utility and use the Collections one.

Reply:
I made the simplification you suggested.